Feature/updating-alignment

thiagogenez opened the pull request:
This pull request deliveries a "plugin" called `cactus-update-prepare` as a wrapper of `cactus-prepare` to provide user-friendly steps to update Cactus alignment.

This plugin provides (at the moment) the update functionality of [adding new genomes](https://github.com/thiagogenez/cactus/blob/master/doc/updating-alignments.md#adding-a-new-genome) to a node.  The functionality of [adding to a branch](https://github.com/thiagogenez/cactus/blob/master/doc/updating-alignments.md#adding-to-a-branch) will come soon. 

Therefore, this plugin furnishes two subcommands:

```
usage: cactus-update-prepare {node,branch} [-h] [Options]

subcommand:
  {node,branch}  Desired alignment update approach
    node         Adding a new genome to a node (aka, the update-node approach)
    branch       Add a new genome to a branch (aka, the update-branch approach)
```

Any feedback is welcome and will be highly appreciated. 

Cheers,
Thiago

glennhickey commented:
Thanks @thiagogenez .  As someone who's recently suffered through updating a big alignment, I can see how useful this is.  Will take a closer look soon.

thiagogenez commented:
> Thanks @thiagogenez. As someone who's recently suffered through updating a big alignment, I can see how useful this is. Will take a closer look soon.

That's great to hear, @glennhickey. Thanks for your comment. 

Yes, I had similar issues while re-rooting an alignment too. It is very easy to make some mistakes while running the HAL-tool kit command lines to make updates.

glennhickey reviewed (changes requested):
I think you need to test this.  Might I suggest adding simChimp from ../examples/evolverPrimates.txt to "mr" from ../examples/evolverMammals.txt ?  

Then you can document the exact command lines run so that I (and future users) can test it out on a small reproducible example.  This could also form the basis of a future automated test. 

Because as it stands, I don't think the tool works at all, at least not in general.  The main issue I see that that if you are adding a new genome to an ancestor in the tree, the ancestor's sequence needs to be fixed.  You can do this with `cactus-align` and `cactus-blast` with the `--includeRoot`  option (making sure that the sequence is extracted from the original hal file).  (see #637 #638),  Without this option, the steps in the documentation (using only `cactus`) work, but I don't see how blast/align could possibly run, as they will infer a new ancestor and halReplaceGenome will throw an exception like
```
halReplaceGenome --bottomAlignmentFile ./steps/out.hal --topAlignmentFile evolverMammals.hal evolverMammals.hal Anc2
terminate called after throwing an instance of 'hal_exception'
  what():  Cannot update sequence Anc2refChr7 because it is not present in  genome Anc2
```

thiagogenez commented on src/cactus/update/cactus_update_prepare.py:
```diff
@@ -0,0 +1,382 @@
+        cactus_options = []
+
+    # cactus_prepare_options is append at the tail of cmd to overide cactus options we pre-defined below
+    cmd = 'cactus-prepare {} --outDir {}/steps --outSeqFile {}/steps/steps.txt --jobStore {}/jobstore {}'.format(
```
Thanks for pointing this out. This is a question I forgot to ask the last meeting about skipping running the preprocessor jobs on extracted genomes from the hal file. I will filter this out.

thiagogenez commented on src/cactus/update/cactus_update_prepare.py:
```diff
@@ -0,0 +1,382 @@
+from tempfile import mkdtemp
+
+# Cactus library
+from cactus.shared.common import cactus_call
```
Thank you for highlighting this. I left the `logging` to be included at the end of this, while `setupBinaries(options)` I forgot to included indeed. Thanks

thiagogenez commented on src/cactus/update/cactus_update_prepare.py:
```diff
@@ -0,0 +1,382 @@
+
+    if genome_internal_id is None:
+        raise RuntimeError(
+            "Genome \'{}\' not found in the tree extracted from the HAL-format file {}".format(options.genome, options.halFile))
```
You are right. This was the old name of the variable pointing to the hal file. Thanks for letting me know.

thiagogenez commented on src/cactus/update/cactus_update_prepare.py:
```diff
@@ -0,0 +1,382 @@
+        sys.exit(1)
+
+    # validate the given inFasta text file
+    options.inFasta = inFasta_sanity_check(options.inFasta)
```
No, the `inFile` is not the usual `SeqFile`. It is just a  list of genomes to be included.  I will rename this variable to avoid further confusion.

I need to improve how the user will provide this piece of information to update its alignment. Any suggestion about this @glennhickey ? From a user point of view, how you would like to provide this piece of information to get your alignment updated?

thiagogenez commented:
Thanks, @glennhickey for your valuable inputs highlighting bugs and improvements for this PR. I will fix them and test on the ` ../examples/evolverPrimates.txt` example as you said.  I will let you know as soon as I have tackled all of them. 

You are right about the `-includeRoot` option. As soon as you point this out I realised that I forgot to take this. I am aware that "`--root`-like" option is enabled on `cactus` but not in the `cactus-prepare`.  I will think about how to cope with this issue.

glennhickey commented:
> You are right about the `-includeRoot` option. As soon as you point this out I realised that I forgot to tackle this carefully. I am aware that "`--root`-like" option is enabled on `cactus` but not in the `cactus-prepare`. I will think about solutions to cope with this issue.

I think the solutions are fairly minor changes: Add the `--includeRoot` option, and make sure to extract the ancestral sequence from the input hal and include it in the input.

thiagogenez commented:
Hi @glennhickey 

I have added [examples](https://github.com/thiagogenez/cactus/tree/feature/updating-alignment/src/cactus/update) and addressed the issues you point out previously. Please let me know your thoughts on this

Next step,
- merging it to your CI tests
- adding `replace` subcommand functionality

glennhickey commented:
This seems to be running now, congrats.  Comments

1. Need to put the documentation somewhere where people can find it.  Suggest moving your README.md into https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/doc/updating-alignments.md.
3. People will run examples from `./cactus` where `./hal` already exists.  Please change your examples to just use `./evolverMammals.hal` and `./input.txt`.  Also, it wouldn't hurt to list the command line required to create evolverMammals.hal
4. It seems to print out unnecessary `hal2fasta` commands after final `cactus-align` in both modes.  Please remove
5. The update branch example in your documentation doesn't work `cactus-update-prepare {node,branch} [-h] [Options] branch: error: the following arguments are required: --parentGenome, --childGenome`
6. The cactus-blast seqfiles when updating branches should contain outgroups.  Yours don't seem to.  Please see https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/doc/updating-alignments.md as well as my test https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/test/evolverTest.py#L119-L164
7. the `input.txt` description seems wrong in add-to-branch section.

thiagogenez commented:
> This seems to be running now, congrats. Comments
> 
> 1. Need to put the documentation somewhere where people can find it.  Suggest moving your README.md into https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/doc/updating-alignments.md.
> 2. People will run examples from `./cactus` where `./hal` already exists.  Please change your examples to just use `./evolverMammals.hal` and `./input.txt`.  Also, it wouldn't hurt to list the command line required to create evolverMammals.hal
> 3. It seems to print out unnecessary `hal2fasta` commands after final `cactus-align` in both modes.  Please remove
> 4. The update branch example in your documentation doesn't work `cactus-update-prepare {node,branch} [-h] [Options] branch: error: the following arguments are required: --parentGenome, --childGenome`
> 5. The cactus-blast seqfiles when updating branches should contain outgroups.  Yours don't seem to.  Please see https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/doc/updating-alignments.md as well as my test https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/test/evolverTest.py#L119-L164
> 6. the `input.txt` description seems wrong in add-to-branch section.

@glennhickey  - Thanks for the feedback above. I'm addressing them. 

Regarding item `5`, how further up in the tree the script should go to capture outgropus? I am asking because my script includes all children of `parentGenome` as outgroups in the first Cactus alignment (following the `genome3` [example](https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/doc/updating-alignments.md) used as outgroup in the first Cactus call)

However, your test above seems to include everybody as an outgroup in the first Cactus call.

Adding `simGorila`(acting as `genome6` in the example) in the branch connecting `mr` (acting as `genome4`) with `Anc1` (acting as `genome5`), my script will create the following tree in the `seqFile`: `(simHuman_chr6:0.144018,(mr:0.171974,simGorilla:0.075)AncGorilla:0.1)Anc1;`, where `simHuman_chr6` (acting as `genome3`) will serve as an outgroup for `AncGorilla` (acting as `genome7`).

glennhickey commented:
Sorry @thiagogenez I somehow never noticed your last message.  I guess (and this applies to both add-to-node and add-to-branch) the only way to do it properly is to let cactus compute the outgroups.  This can be done fairly simply with this function (run on the full tree + nodes you want to add): 
https://github.com/ComparativeGenomicsToolkit/cactus/blob/paf_chains/src/cactus/progressive/progressive_decomposition.py#L53
But unfortunately, the equivalent logic in the master branch is way more convoluted.  

So, I think the easiest thing to do would be to rebase your PR on the paf_chains branch and work from there.  Is that okay? It'll be a few weeks before it gets into master, but it will be in the next release.

thiagogenez commented:
> Sorry @thiagogenez I somehow never noticed your last message. I guess (and this applies to both add-to-node and add-to-branch) the only way to do it properly is to let cactus compute the outgroups. This can be done fairly simply with this function (run on the full tree + nodes you want to add): https://github.com/ComparativeGenomicsToolkit/cactus/blob/paf_chains/src/cactus/progressive/progressive_decomposition.py#L53 But unfortunately, the equivalent logic in the master branch is way more convoluted.
> 
> So, I think the easiest thing to do would be to rebase your PR on the paf_chains branch and work from there. Is that okay? It'll be a few weeks before it gets into master, but it will be in the next release.

I see - thank you, @glennhickey, for the reply. No problem, I will rebase my PR to use the `paf_chains`'s `compute_outgroups` function then. Thanks

dthybert commented:
> 

I better understand now the issue. I do agree we need an out-group to infer genome 7 (and I thought that Thiago was using an out-group for this). Selecting anc5 would be the easiest for us as we would know in advance which assembly to use.

glennhickey commented:
I just reran @thiago's code, and "human" which is the equivalent of "genome3" in the example does appear.  So this also makes my original comment in error.  Sorry again.
>The cactus-blast seqfiles when updating branches should contain outgroups. Yours don't seem to

Could could "genome 7" in the example benefit from additional outgroups beyond "Genome 3" (if they exist, there's no room for them on the small example tree)? I think it's possible, which would be why letting cactus choose 3 outgroups would might be better.  

Choosing "genome 5" is possible, but note that it would have to be set as a leaf in the "genome 7" alignment tree (which is easy enough by rerooting it).

thiagogenez commented:
Hi @glennhickey and @dthybert 
Thanks for your valuable input and comments on this matter. I have refactored a few bits and improved the documentation.  Sorry for the misunderstanding regarding the `outgroup` topic and I believe it is ready the code is ready to merge.  

Please, bear in mind that:

- This script creates a Newick tree following the updating alignments [rules](https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/doc/updating-alignments.md) for both `adding-to-a-node` and `adding-to-branch` cases. The tree is stored in the file called `seq_file.in` and used as an input by `cactus-prepare` (which then generates the `seq_file.out` file).
> - This means that, for the `adding-to-a-branch` case discussed above, the `genome3` is embedded in the `seqFile`'s Newick tree as a child of `genome5` and sibling of `genome7`. 

- For the `adding-to-a-branch` case, the script does _not_ put any effort into computing a set of nodes to be used as an outgroup. Instead, it only attempts to extract as many FASTA files from the given HAL file as needed for the two `Cactus` calls. Nodes (potential/candidate outgroups) above the one represented by `--parentGenome` are not analysed by the script, and therefore no FASTA file is extracted.
> - This means that, for the `adding-to-a-branch` case discussed above, the FASTA files of  `genome5` and its children (`genome3`), as well as `genome4` and `genome6`, are extracted from the HAL file.

glennhickey commented:
Thanks, @thiagogenez!!
